# Music Streaming Wars: Song Popularity Prediction

Author: E. Berke Tezcan

<img src="./images/apple-music-spotify-tidal.jpg" style="width: 60%">

***

## TABLE OF CONTENTS 

*Click to jump to matching Markdown Header.*<br><br>
 
- **[Introduction](#INTRODUCTION)<br>**
- **[OBTAIN](#OBTAIN)**<br>
- **[SCRUB/EXPLORE](#SCRUB/EXPLORE)**<br>
- **[MODEL](#MODEL)**<br>
- **[iNTERPRET](#iNTERPRET)**<br>
- **[Conclusions/Recommendations](#CONCLUSIONS-&-RECOMMENDATIONS)<br>**
___

# INTRODUCTION

With Apple Music announcing on May 17th that they will be providing lossless audio along with spatial audio by Dolby Atmos for their subscribers and Tidal continuously providing exclusive content from artists, the competition among audio streaming platforms is heating up. Spotify would like to stay competitive by being able to predict which songs are going to be popular ahead of time so that they can curate even better playlists and sign deals with up-and-coming artists to have exclusivity on their content. This would not only help retain the current subscribers but also help market the platform to new subscribers as well.

For this project, we were hired by Spotify to develop a machine learning model that can accurately predict whether a song is going to be popular or not. In order to achieve this, we will be evaluating different machine learning models and will look at what attributes of a song are the most important for determining its popularity.

# OBTAIN

We will be using a dataset from Kaggle (https://www.kaggle.com/zaheenhamidani/ultimate-spotify-tracks-db) that contains approximately 232,000 tracks and their attributes to train several machine learning models in order to find the common threads between popular songs.

In [1]:
import pandas as pd

In [2]:
#importing data into a dataframe
df = pd.read_csv('./data/SpotifyFeatures.csv')
df.head()

genre        artist_name                        track_name  \
0  Movie     Henri Salvador       C'est beau de faire un Show   
1  Movie  Martin & les fées  Perdu d'avance (par Gad Elmaleh)   
2  Movie    Joseph Williams    Don't Let Me Be Lonely Tonight   
3  Movie     Henri Salvador    Dis-moi Monsieur Gordon Cooper   
4  Movie       Fabien Nataf                         Ouverture   

                 track_id  popularity  acousticness  danceability  \
0  0BRjO6ga9RKCKjfDqeFgWV           0         0.611         0.389   
1  0BjC1NfoEOOusryehmNudP           1         0.246         0.590   
2  0CoSDzoNIKCRs124s9uTVy           3         0.952         0.663   
3  0Gc6TVm52BwZD07Ki6tIvf           0         0.703         0.240   
4  0IuslXpMROHdEPvSl1fTQK           4         0.950         0.331   

   duration_ms  energy  instrumentalness key  liveness  loudness   mode  \
0        99373   0.910             0.000  C#    0.3460    -1.828  Major   
1       137373   0.737             0.000  F#    0.1510    -5.559  Minor   
2       170267   0.131             0.000   C    0.1030   -13.879  Minor   
3       152427   0.326             0.000  C#    0.0985   -12.178  Major   
4        82625   0.225             0.123   F    0.2020   -21.150  Major   

   speechiness    tempo time_signature  valence  
0       0.0525  166.969            4/4    0.814  
1       0.0868  174.003            4/4    0.816  
2       0.0362   99.488            5/4    0.368  
3       0.0395  171.758            4/4    0.227  
4       0.0456  140.576            4/4    0.390

In [3]:
#looking at the stats of different columns
df.describe()

popularity   acousticness   danceability   duration_ms  \
count  232725.000000  232725.000000  232725.000000  2.327250e+05   
mean       41.127502       0.368560       0.554364  2.351223e+05   
std        18.189948       0.354768       0.185608  1.189359e+05   
min         0.000000       0.000000       0.056900  1.538700e+04   
25%        29.000000       0.037600       0.435000  1.828570e+05   
50%        43.000000       0.232000       0.571000  2.204270e+05   
75%        55.000000       0.722000       0.692000  2.657680e+05   
max       100.000000       0.996000       0.989000  5.552917e+06   

              energy  instrumentalness       liveness       loudness  \
count  232725.000000     232725.000000  232725.000000  232725.000000   
mean        0.570958          0.148301       0.215009      -9.569885   
std         0.263456          0.302768       0.198273       5.998204   
min         0.000020          0.000000       0.009670     -52.457000   
25%         0.385000          0.000000       0.097400     -11.771000   
50%         0.605000          0.000044       0.128000      -7.762000   
75%         0.787000          0.035800       0.264000      -5.501000   
max         0.999000          0.999000       1.000000       3.744000   

         speechiness          tempo        valence  
count  232725.000000  232725.000000  232725.000000  
mean        0.120765     117.666585       0.454917  
std         0.185518      30.898907       0.260065  
min         0.022200      30.379000       0.000000  
25%         0.036700      92.959000       0.237000  
50%         0.050100     115.778000       0.444000  
75%         0.105000     139.054000       0.660000  
max         0.967000     242.903000       1.000000

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 232725 entries, 0 to 232724
Data columns (total 18 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   genre             232725 non-null  object 
 1   artist_name       232725 non-null  object 
 2   track_name        232725 non-null  object 
 3   track_id          232725 non-null  object 
 4   popularity        232725 non-null  int64  
 5   acousticness      232725 non-null  float64
 6   danceability      232725 non-null  float64
 7   duration_ms       232725 non-null  int64  
 8   energy            232725 non-null  float64
 9   instrumentalness  232725 non-null  float64
 10  key               232725 non-null  object 
 11  liveness          232725 non-null  float64
 12  loudness          232725 non-null  float64
 13  mode              232725 non-null  object 
 14  speechiness       232725 non-null  float64
 15  tempo             232725 non-null  float64
 16  time_signature    23

We once again see that we have 232,725 tracks in the dataset with both categorical and numerical columns. In order to use the information from the categorical columns ('genre', 'artist_name', 'track_name', 'track_id', 'key', 'mode', 'time_signature') we will either need to represent them numerically by feature engineering or drop them to be able to train the models.

In [5]:
#looking at different values contained within columns
for col in df.columns:
    print(f"Column: {col}")
    print(df[col].value_counts())
    print("--------------------")

Column: genre
Comedy              9681
Soundtrack          9646
Indie               9543
Jazz                9441
Pop                 9386
Electronic          9377
Children’s Music    9353
Folk                9299
Hip-Hop             9295
Rock                9272
Alternative         9263
Classical           9256
Rap                 9232
World               9096
Soul                9089
Blues               9023
R&B                 8992
Anime               8936
Reggaeton           8927
Ska                 8874
Reggae              8771
Dance               8701
Country             8664
Opera               8280
Movie               7806
Children's Music    5403
A Capella            119
Name: genre, dtype: int64
--------------------
Column: artist_name
Giuseppe Verdi               1394
Giacomo Puccini              1137
Kimbo Children's Music        971
Nobuo Uematsu                 825
Richard Wagner                804
                             ... 
Grits                           1
Marcus

There are a couple things that stand out in the value counts of the columns. First one is that we have the "Children's Music" genre showing up twice and we have duplicated values in the track_id column.

# SCRUB/EXPLORE

## Addressing "Children's Music" Character Discrepancy

In [6]:
df['genre'].value_counts()

Comedy              9681
Soundtrack          9646
Indie               9543
Jazz                9441
Pop                 9386
Electronic          9377
Children’s Music    9353
Folk                9299
Hip-Hop             9295
Rock                9272
Alternative         9263
Classical           9256
Rap                 9232
World               9096
Soul                9089
Blues               9023
R&B                 8992
Anime               8936
Reggaeton           8927
Ska                 8874
Reggae              8771
Dance               8701
Country             8664
Opera               8280
Movie               7806
Children's Music    5403
A Capella            119
Name: genre, dtype: int64

There are 2 types of "Children's Music" values in the genres due to the character used for apostrophe. Since both of these values are meant to show the same thing we need to merge them and achieve consistency.

In [7]:
df.loc[df['genre']=="Children’s Music",'genre']="Children's Music"

In [8]:
#verifying that the issue has been resolved
df['genre'].value_counts()

Children's Music    14756
Comedy               9681
Soundtrack           9646
Indie                9543
Jazz                 9441
Pop                  9386
Electronic           9377
Folk                 9299
Hip-Hop              9295
Rock                 9272
Alternative          9263
Classical            9256
Rap                  9232
World                9096
Soul                 9089
Blues                9023
R&B                  8992
Anime                8936
Reggaeton            8927
Ska                  8874
Reggae               8771
Dance                8701
Country              8664
Opera                8280
Movie                7806
A Capella             119
Name: genre, dtype: int64

## Missing Values

In [9]:
#checking for missing values
df.isna().sum()

genre               0
artist_name         0
track_name          0
track_id            0
popularity          0
acousticness        0
danceability        0
duration_ms         0
energy              0
instrumentalness    0
key                 0
liveness            0
loudness            0
mode                0
speechiness         0
tempo               0
time_signature      0
valence             0
dtype: int64

We don't have any missing values in our columns so we will move onto check for duplicated rows.

## Addressing Duplicated Tracks

We need to take a look and find all duplicated tracks by using their unique id numbers.

In [10]:
df[df['track_id'].duplicated()]

genre   artist_name  \
1348    Alternative      Doja Cat   
1385    Alternative   Frank Ocean   
1452    Alternative   Frank Ocean   
1554    Alternative    Steve Lacy   
1634    Alternative      tobi lou   
...             ...           ...   
232715         Soul    Emily King   
232718         Soul  Muddy Waters   
232720         Soul         Slave   
232722         Soul  Muddy Waters   
232723         Soul       R.LUM.R   

                                               track_name  \
1348                                           Go To Town   
1385                                            Seigfried   
1452                                         Bad Religion   
1554                                                 Some   
1634                                            Buff Baby   
...                                                   ...   
232715                                               Down   
232718  I Just Want To Make Love To You - Electric Mud...   
232720                                       Son Of Slide   
232722                     (I'm Your) Hoochie Coochie Man   
232723                                      With My Words   

                      track_id  popularity  acousticness  danceability  \
1348    6iOvnACn4ChlAw4lWUU4dd          64       0.07160         0.710   
1385    1BViPjTT585XAhkUUrkts0          61       0.97500         0.377   
1452    2pMPWE7PJH1PizfgGRMnR9          56       0.77900         0.276   
1554    4riDfclV7kPDT9D58FpmHd          58       0.00548         0.784   
1634    1F1QmI8TMHir9SUFrooq5F          59       0.19000         0.736   
...                        ...         ...           ...           ...   
232715  5cA0vB8c9FMOVDWyJHgf26          42       0.55000         0.394   
232718  2HFczeynfKGiM9KF2z2K7K          43       0.01360         0.294   
232720  2XGLdVl7lGeq8ksM6Al7jT          39       0.00384         0.687   
232722  2ziWXUmQLrXTiYjCg2fZ2t          47       0.90100         0.517   
232723  6EFsue2YbIG4Qkq8Zr9Rir          44       0.26200         0.745   

        duration_ms  energy  instrumentalness key  liveness  loudness   mode  \
1348         217813   0.710          0.000001   C    0.2060    -2.474  Major   
1385         334570   0.255          0.000208   E    0.1020   -11.165  Minor   
1452         175453   0.358          0.000003   A    0.0728    -7.684  Major   
1554         118393   0.554          0.254000   G    0.0995    -6.417  Major   
1634         215385   0.643          0.000000   F    0.1060    -8.636  Major   
...             ...     ...               ...  ..       ...       ...    ...   
232715       281853   0.346          0.000002   E    0.1290   -13.617  Major   
232718       258267   0.739          0.004820   C    0.1380    -7.167  Major   
232720       326240   0.714          0.544000   D    0.0845   -10.626  Major   
232722       166960   0.419          0.000000   D    0.0945    -8.282  Major   
232723       222442   0.704          0.000000   A    0.3330    -7.137  Major   

        speechiness    tempo time_signature  valence  
1348         0.0579  169.944            4/4    0.700  
1385         0.0387  125.004            5/4    0.370  
1452         0.0443   81.977            4/4    0.130  
1554         0.0300  104.010            4/4    0.634  
1634         0.0461  156.002            4/4    0.599  
...             ...      ...            ...      ...  
232715       0.0635   90.831            4/4    0.436  
232718       0.0434  176.402            4/4    0.945  
232720       0.0316  115.542            4/4    0.962  
232722       0.1480   84.135            4/4    0.813  
232723       0.1460  100.031            4/4    0.489  

[55951 rows x 18 columns]

We have 55,951 duplicated rows that we need to address. Before we can address these duplications though we need to see what the cause of the duplicates are.

In [11]:
#checking rows for duplicated ids to see differences
df[df['track_id']=='6iOvnACn4ChlAw4lWUU4dd'] 

genre artist_name  track_name                track_id  \
257                  R&B    Doja Cat  Go To Town  6iOvnACn4ChlAw4lWUU4dd   
1348         Alternative    Doja Cat  Go To Town  6iOvnACn4ChlAw4lWUU4dd   
77710   Children's Music    Doja Cat  Go To Town  6iOvnACn4ChlAw4lWUU4dd   
93651              Indie    Doja Cat  Go To Town  6iOvnACn4ChlAw4lWUU4dd   
113770               Pop    Doja Cat  Go To Town  6iOvnACn4ChlAw4lWUU4dd   

        popularity  acousticness  danceability  duration_ms  energy  \
257             64        0.0716          0.71       217813    0.71   
1348            64        0.0716          0.71       217813    0.71   
77710           64        0.0716          0.71       217813    0.71   
93651           64        0.0716          0.71       217813    0.71   
113770          64        0.0716          0.71       217813    0.71   

        instrumentalness key  liveness  loudness   mode  speechiness    tempo  \
257             0.000001   C     0.206    -2.474  Major       0.0579  169.944   
1348            0.000001   C     0.206    -2.474  Major       0.0579  169.944   
77710           0.000001   C     0.206    -2.474  Major       0.0579  169.944   
93651           0.000001   C     0.206    -2.474  Major       0.0579  169.944   
113770          0.000001   C     0.206    -2.474  Major       0.0579  169.944   

       time_signature  valence  
257               4/4      0.7  
1348              4/4      0.7  
77710             4/4      0.7  
93651             4/4      0.7  
113770            4/4      0.7

In [12]:
df[df['track_id']=='2XGLdVl7lGeq8ksM6Al7jT']

genre artist_name    track_name                track_id  popularity  \
179212  Jazz       Slave  Son Of Slide  2XGLdVl7lGeq8ksM6Al7jT          39   
232720  Soul       Slave  Son Of Slide  2XGLdVl7lGeq8ksM6Al7jT          39   

        acousticness  danceability  duration_ms  energy  instrumentalness key  \
179212       0.00384         0.687       326240   0.714             0.544   D   
232720       0.00384         0.687       326240   0.714             0.544   D   

        liveness  loudness   mode  speechiness    tempo time_signature  \
179212    0.0845   -10.626  Major       0.0316  115.542            4/4   
232720    0.0845   -10.626  Major       0.0316  115.542            4/4   

        valence  
179212    0.962  
232720    0.962

In [13]:
df[df['track_id']=='2HFczeynfKGiM9KF2z2K7K']


genre   artist_name  \
48555   Blues  Muddy Waters   
232718   Soul  Muddy Waters   

                                               track_name  \
48555   I Just Want To Make Love To You - Electric Mud...   
232718  I Just Want To Make Love To You - Electric Mud...   

                      track_id  popularity  acousticness  danceability  \
48555   2HFczeynfKGiM9KF2z2K7K          35        0.0136         0.294   
232718  2HFczeynfKGiM9KF2z2K7K          43        0.0136         0.294   

        duration_ms  energy  instrumentalness key  liveness  loudness   mode  \
48555        258267   0.739           0.00482   C     0.138    -7.167  Major   
232718       258267   0.739           0.00482   C     0.138    -7.167  Major   

        speechiness    tempo time_signature  valence  
48555        0.0434  176.402            4/4    0.945  
232718       0.0434  176.402            4/4    0.945

We see that most of the attributes of the duplicated songs are the same except for 'popularity' and 'genre'. The 'popularity' column can be aggregated since it is a numerical column but the categorical column of 'genre' is a little bit trickier. What makes the most sense in this case would be to create different columns with the genre names and display with binary values whether a song belongs to that genre or not.

In [14]:
#generating a list with the genre names
genre_list = list(df['genre'].unique())

In [15]:
#creating the genre columns using the genre list
for genre in genre_list:
    df[genre] = (df['genre']==genre).astype('int')

In [ ]:
#grouping by track_id number to get rid of duplicates and keeping the maximum values in each column.
df=df.groupby(['track_id']).max()

Above, we created the genre columns and merged the duplicated values keeping the maximum value in each column. This makes sense since the track that is being listened to is the same one. For example, if a track had popularity scores of 15, 25, 38 and 42 in its duplicated rows, we are keeping the best value of 42 by taking the max.

In [ ]:
#removing redundant genre column
df.drop('genre', axis=1, inplace=True)
df.head()

In [ ]:
#verifying that duplicates have been eliminated
df[df.index =='6iOvnACn4ChlAw4lWUU4dd']

We successfully addressed the duplicates of each track by aggregating them to a single row.

In [ ]:
df.info()

We now have 176,774 unique tracks in our dataset (down from 232,725).

## Feature Engineering - is_popular

Since our goal is to be able to identify which tracks will be popular, we need to feature engineer a new column by binarizing the popularity column. To be able to do this, we need to decide on a cut-off point of popularity score which if a song stays above this cut-off point it will be considered "popular" and if it stays below it will be considered "not popular". We can start off by taking a look at the distribution of the popularity score distribution.

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
#creating a histogram to see distribution of popularity scores in the dataset.
sns.histplot(df['popularity'], bins='auto')

From the above histogram we see that we have a bimodal distribution. One of the peaks is at 0, and the other one seems to be around 40. In order to better decide what's popular, we can take a look at the Top 50 songs' popularity scores (this data is also from 2019 similar to our main dataset to keep the analysis consistent.)

### Top 50 Songs - 2019

In [ ]:
#data from https://www.kaggle.com/leonardopena/top50spotify2019
df_50 = pd.read_csv('data/top50.csv', encoding='latin1', index_col=0)

In [ ]:
df_50.head()

In [ ]:
#displaying stats information of Top 50 songs
df_50['Popularity'].describe()

Going back to our histogram we can draw vertical lines to see where these values fall into.

In [ ]:
fig, ax = plt.subplots()
sns.histplot(df['popularity'], bins='auto', ax=ax)
stats=['mean', '50%', 'min', 'max']
for stat in stats:
    ax.vlines(x=df_50['Popularity'].describe()[stat], ymin=0, ymax=6000, linestyles='dashed', colors='red', label=stat)

We can see that there was a range of popularity scores in the Top 50 songs between 70 and 95. Which means that any song that is above a 70 theoretically could be a popular song. It doesn't make sense to use median or mean scores for our cutoff point in this case since then we would be disregarding all the songs that had lower values than 87.5 or 88 as unpopular which is untrue. However, before we can establish the cutoff point we need to acknowledge that we are basing it off of only 50 datapoints which is not a lot. It may be good to take a look at Top 100 songs instead of 50 to get a better sample size of popular songs.

### Top 100 Songs - 2019

In [ ]:
#data from https://www.kaggle.com/reach2ashish/top-100-spotify-songs-2019
df_100 = pd.read_csv('data/spotify_top_100_2019.csv')

In [ ]:
df_100['popularity '].describe()

The minimum value of 4 for the popularity score on the Top 100 Songs chart seems like an outlier. Next, we'll visualize the spread of this column to confirm.

In [ ]:
fig, ax = plt.subplots()
sns.histplot(df_100['popularity '], bins='auto', ax=ax)

As we imagined the scores within the range 0-25 seem like outliers. We can remove outliers from this dataset with the IQR method to get a better perspective on the data.

In [ ]:
#Outlier Removal with the IQR method

def find_outliers_IQR(data):
    """Use Tukey's Method of outlier removal AKA InterQuartile-Range Rule
    and return boolean series where True indicates it is an outlier.
    - Calculates the range between the 75% and 25% quartiles
    - Outliers fall outside upper and lower limits, using a treshold of  1.5*IQR the 75% and 25% quartiles.

    IQR Range Calculation:    
        res = df.describe()
        IQR = res['75%'] -  res['25%']
        lower_limit = res['25%'] - 1.5*IQR
        upper_limit = res['75%'] + 1.5*IQR

    Args:
        data (Series,or ndarray): data to test for outliers.

    Returns:
        [boolean Series]: A True/False for each row use to slice outliers.
        
    EXAMPLE USE: 
    >> idx_outs = find_outliers_df(df['AdjustedCompensation'])
    >> good_data = df[~idx_outs].copy()
    
    function snippet from Flatiron School Phase #2 Py Files.
    URL = https://github.com/flatiron-school/Online-DS-FT-022221-Cohort-Notes/blob/master/py_files/functions_SG.py
    
    """
    df_b=data
    res= df_b.describe()

    IQR = res['75%'] -  res['25%']
    lower_limit = res['25%'] - 1.5*IQR
    upper_limit = res['75%'] + 1.5*IQR

    idx_outs = (df_b>upper_limit) | (df_b<lower_limit)

    return idx_outs

In [ ]:
#removing outliers from the popularity column
df_100 = df_100[find_outliers_IQR(df_100['popularity '])==False]
#displaying minimum & maxium values in popularity column 
print("Minimum:", df_100['popularity '].min())
print("Maximum:", df_100['popularity '].max())

In [ ]:
fig, ax = plt.subplots()
sns.histplot(df_100['popularity '], bins=15, ax=ax)

In [ ]:
#visualizing the min and max popularity scores on the overall dataset histogram
fig, ax = plt.subplots()
sns.histplot(df['popularity'], bins='auto', ax=ax)
ax.vlines(x=df_100['popularity '].min(), ymin=0, ymax=6000, linestyles='dashed', colors='red', label='min')
ax.vlines(x=df_100['popularity '].max(), ymin=0, ymax=6000, linestyles='dashed', colors='red', label='max')
plt.legend()

As we can expect to see, the top 100 songs have a wider range and therefore a lower popularity score threshold compared to the top 50 songs. We will be defining a song being popular as being Top 100 worthy and therefore will establish our cutoff point at 58.

In [ ]:
#creating is_popular column with our cutoff point
df['is_popular']=(df['popularity']>=58).astype('int')
df.head()

In [ ]:
#dropping popularity score column since we will not be using it
df.drop(['popularity', 'artist_name', 'track_name'], axis=1, inplace=True)
df.head()

We dropped popularity scores since we already binarized that column, but additionally we are dropping 'artist_name' and 'track_name' since we are looking at the anatomy of a song and not who sings it or what it's called. The goal is to identify songs that will become popular without being affected by the artist's name since we would also like to find songs from up-and-coming artists.

## One Hot Encoding the Categorical Columns

We still have categorical columns that need one hot encoding. Namely, these columns are 'key', 'mode' and 'time_signature'.

In [ ]:
#Check to see how many more columns we will be creating by OHE the cat_cols.
df.nunique()

We will be creating 2 (mode) + 5 (time_signature) + key (12) - 3 (drop_first) = 16 columns.

In [ ]:
#define categorical columns
cat_cols = ['key', 'mode', 'time_signature']

In [ ]:
#One hot encoding the dataframe
from sklearn.preprocessing import OneHotEncoder

encoder=OneHotEncoder(sparse=False, drop='first')
data_ohe = encoder.fit_transform(df[cat_cols])
df_ohe = pd.DataFrame(data_ohe, columns=encoder.get_feature_names(cat_cols), index=df.index)

In [ ]:
pd.set_option("display.max_columns", None)
df_ohe

In [ ]:
df_ohe = pd.concat([df.drop(cat_cols, axis=1), df_ohe], axis=1)
df_ohe.head()

With the dataframe scrubbed and one hot encoded we can move onto the modelling process.

# MODEL

## train_test_split

In [ ]:
#splitting the data to training and test sets in order to be able to measure performance
from sklearn.model_selection import train_test_split
y=df_ohe['is_popular']
X=df_ohe.drop('is_popular',axis=1)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=42)

The first model we will be generating is a dummy classifier. We will be comparing our models' success to each other but also to this baseline model.

## Model #1 - Baseline - Dummy Classifier

In [ ]:
from sklearn.dummy import DummyClassifier

clf_dummy = DummyClassifier(random_state=42)
clf_dummy.fit(X_train, y_train)
y_pred = clf_dummy.predict(X_test)

We need a function that will show us the classification report, the confusion matrix as well as the ROC curve to be able to evaluate our models.

In [ ]:
from sklearn.metrics import classification_report, plot_confusion_matrix, plot_roc_curve

def classification(y_true, y_pred, X, clf):
    """This function shows the classification report,
    the confusion matrix as well as the ROC curve for evaluation of model quality.
    
    y_true: Correct y values, typically y_test that comes from the train_test_split performed at the beginning of model development.
    y_pred: Predicted y values by the model.
    clf: classifier model that was fit to training data.
    X: X_test values"""
    
    #Classification report
    print("CLASSIFICATION REPORT")
    print("------------------------------------------")
    print(classification_report(y_true=y_true, y_pred=y_pred))
    
    #Creating a figure/axes for confusion matrix and ROC curve
    fig, ax = plt.subplots(ncols=2, figsize=(12, 5))
    
    #Plotting the normalized confusion matrix
    plot_confusion_matrix(estimator=clf, X=X, y_true=y_true, cmap='Blues', normalize='true', ax=ax[0])
    
    #Plotting the ROC curve
    plot_roc_curve(estimator=clf, X=X, y=y_true, ax=ax[1])
    
    #Plotting the 50-50 guessing plot for reference
    ax[1].plot([0,1], [0,1], ls='--', color='orange')

In [ ]:
classification(y_test, y_pred, X_test, clf_dummy)

In [ ]:
#class imbalance percentages
y_train.value_counts(normalize=True)

Our dummy classifier correctly predicted 89% of the unpopular songs as unpopular; however, it correctly predicted only 11% of the popular songs as popular and instead classified 89% of them as unpopular as well. We clearly have a class imbalance problem where approximately 88% of our data is not popular and only about 11% of it is. To address this we can SMOTE the training data and see if training a model with this method would improve our results.

## Addressing Class Imbalance with SMOTENC

In [ ]:
#looking at column names to extract categorical column indices for SMOTENC
X.columns

In [ ]:
#creating a list of categorical column indices
cat_cols = list(range(10, len(X.columns)))
X.columns[cat_cols]

In [ ]:
#Using SMOTENC to address class imbalance. We are not using SMOTE since we have categorical columns.
from imblearn.over_sampling import SMOTE, SMOTENC

sm = SMOTENC(categorical_features=cat_cols, random_state=42)

X_train_sm, y_train_sm = sm.fit_resample(X_train, y_train)
y_train_sm.value_counts(normalize=True)

Now that we addressed our class imbalance problem, we can look at the performance of the dummy classifier model once again to use as our baseline.

In [ ]:
#fitting Dummy Classifier to data without the class imbalance problem to serve as a true baseline
clf_dummy_sm = DummyClassifier(random_state=42)
clf_dummy_sm.fit(X_train_sm, y_train_sm)
y_pred = clf_dummy_sm.predict(X_test)
classification(y_test, y_pred, X_test, clf_dummy_sm)

We see here that the dummy classifier is essentially flipping a coin and guessing whether a song is popular or not which is not very useful. However, this serves as a great baseline for our other models to be evaluated against.

## Model #2 - Random Forest Classifier

The first model we will be developing is the Random Forest classifier. 

### Initial Model

In [ ]:
#Fitting RF Classifier to SMOTE'd data
from sklearn.ensemble import RandomForestClassifier

clf_rf = RandomForestClassifier(random_state=42)
clf_rf.fit(X_train_sm, y_train_sm)

#Making predictions and evaluation.
y_pred = clf_rf.predict(X_test)
classification(y_test, y_pred, X_test, clf_rf)

Our Random Forest model performs 48% better than the baseline classifier in predicting unpopular songs correctly and 9% better in predicting popular songs. The model may be overfitting, so to confirm we will look at the performance of the model with the training data.

In [ ]:
#Evaluating the model performance for the training data
y_pred = clf_rf.predict(X_train_sm)
classification(y_train_sm, y_pred, X_train_sm, clf_rf)

Our model is performing perfectly on the training data but not so much on the test data since it is overfitting to the training set. We need to tune our model to get more accurate results on unseen data. We will be using a grid search to optimize for the recall score. We are optimizing recall instead of other scores since we primarily care about correctly identifying a song that will be popular and we don't mind it if we pick a few songs that don't end up becoming popular.

### Hyperparameter Tuning

In [ ]:
# from sklearn.model_selection import GridSearchCV

# clf = RandomForestClassifier()
# grid = {'criterion': ['gini', 'entropy'], 
#         'max_depth': [10, 20, None],
#         'min_samples_leaf': [1, 2, 3]
#        }

# gridsearch = GridSearchCV(estimator=clf, param_grid = grid, scoring='recall')

# gridsearch.fit(X_train_sm,  y_train_sm)
# gridsearch.best_params_
# #Results: {'criterion': 'entropy', 'max_depth': None, 'min_samples_leaf': 2}

In [ ]:
clf_rf_tuned = RandomForestClassifier(criterion='entropy', max_depth=None, 
                                      min_samples_leaf=2, class_weight='balanced', 
                                      random_state=42)
clf_rf_tuned.fit(X_train_sm, y_train_sm)

y_pred = clf_rf_tuned.predict(X_test)
classification(y_test, y_pred, X_test, clf_rf_tuned)

Tuning the hyperparameters of our model improved the recall score for predicting popular songs by 1% (refer to Limitations & Next Steps section on the README document for more information). We can proceed with trying additional types of models to see if the recall score improves.

## Model #3 - XGBoost

### Initial Model

In [ ]:
#Fitting XGBoost classifier to training data and evaluating results
from xgboost import XGBClassifier

clf_xgb = XGBClassifier(random_state=42)
clf_xgb.fit(X_train_sm, y_train_sm)
y_pred = clf_xgb.predict(X_test)
classification(y_test, y_pred, X_test, clf_xgb)

The XGBoost model performed 13% better than the baseline model and 4% better than the random forest model in predicting popular songs right out of the box. We can see how it performs on the training data to see whether it is overfitting and try to tune it if it is.

In [ ]:
#Evaluating the model performance for the training data
y_pred = clf_xgb.predict(X_train_sm)
classification(y_train_sm, y_pred, X_train_sm, clf_xgb)

Once again, we see here that our model is overfitting the training data. We can run another gridsearch and tune our model to see if the recall score can be improved.

### Hyperparameter Tuning

In [ ]:
# grid = {
#     'learning_rate': [0.01, 0.1, 0.2],
#     'max_depth': [10, 20, None]
#         }
# gridsearch = GridSearchCV(estimator=clf_xgb, param_grid = grid, scoring='recall', n_jobs=-1, verbose=2)

# gridsearch.fit(X_train_sm,  y_train_sm)
# gridsearch.best_params_
# # Results: {'learning_rate': 0.1, 'max_depth': 10}

In [ ]:
clf_xgb_tuned = XGBClassifier(learning_rate=0.1, max_depth=10, 
                              random_state=42)
clf_xgb_tuned.fit(X_train_sm, y_train_sm)
y_pred = clf_xgb_tuned.predict(X_test)
classification(y_test, y_pred, X_test, clf_xgb_tuned)

Tuning our model has led to an increase of performance in our recall score by 1%, so we are performing 14% better compared to our baseline Dummy Classifier model and 5% better than our tuned Random Forest model. Next we will try a logistic regression model.

## Model #4 - LogisticRegressionCV

Since the Logistic Regression models are potentially sensitive to outliers and need scaled data we will need to process our data one more time to remove outliers and scale it.

### Removing Outliers

In [ ]:
#separating out the numerical columns for outlier removal
num_cols = list(X.columns[0:10])
num_cols

In [ ]:
#copying df for outlier removal
df_ohe_clean = df_ohe.copy()

In [ ]:
for col in num_cols:
    df_ohe_clean = df_ohe_clean[find_outliers_IQR(df_ohe_clean[col])==False]
    
df_ohe_clean

### train_test_split 

In [ ]:
y=df_ohe_clean['is_popular']
X=df_ohe_clean.drop('is_popular', axis=1)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.30, random_state=42)

### Addressing Class Imbalance with SMOTENC

In [ ]:
y_train.value_counts(normalize=True)

Once again our data has a class imbalance issue so we will be using SMOTENC to address this.

In [ ]:
X_train.columns

In [ ]:
cat_cols = list(range(10,len(X_train.columns)))

In [ ]:
sm = SMOTENC(categorical_features=cat_cols, random_state=42)

X_train_sm, y_train_sm = sm.fit_resample(X_train, y_train)
y_train_sm.value_counts(normalize=True)

### Scaling the Data

In [ ]:
#Using Standard Scaler to scale the smote'd data
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train_sm_sc = scaler.fit_transform(X_train_sm)
X_test_sc = scaler.transform(X_test)

In [ ]:
from sklearn.linear_model import LogisticRegressionCV
clf_logregcv = LogisticRegressionCV(cv=5, random_state=42)
clf_logregcv.fit(X_train_sm_sc,  y_train_sm)
y_pred = clf_logregcv.predict(X_test_sc)
classification(y_test, y_pred, X_test_sc, clf_logregcv)

The logistic regression model has the same recall as our tuned XGBoost model for predicting popular songs while the recall score for predicting unpopular songs is 4% lower. Once again, we will check to see if the model is overfitting and tune the model if it is.

In [ ]:
#Evaluating the model performance for the training data
y_pred = clf_logregcv.predict(X_train_sm_sc)
classification(y_train_sm, y_pred, X_train_sm_sc, clf_logregcv)

Our model is once again overfitting to the training data and performing very well on it but the model's performance drops significantly when we test it with the test data. In order to address this, we can once again perform a grid search and try to tune the model.

### Hyperparameter Tuning

In [ ]:
# clf = LogisticRegressionCV(cv=5)
# grid = {'penalty': ['l1','l2'],
#         'solver': ['liblinear', 'lbfgs', 'sag', 'saga'],
#         'class_weight': ['balanced', None],
#         'Cs': [1e12, 10, 1, 0.1]
#    }

# gridsearch = GridSearchCV(estimator=clf, param_grid = grid, scoring='recall', n_jobs=-1)

# gridsearch.fit(X_train_sm_sc,  y_train_sm)
# gridsearch.best_params_
# # {'Cs': 1, 'class_weight': 'balanced', 'penalty': 'l2', 'solver': 'liblinear'}

The grid search returned 'l2' as the regularization method which is the Ridge regularization as well as a C value of 1. We will use these parameters on a new model to see if the recall score improves.

In [ ]:
clf_logregcv_tuned = LogisticRegressionCV(cv=5, class_weight='balanced', Cs=1, 
                                          penalty='l2', solver='liblinear',
                                          random_state=42)
clf_logregcv_tuned.fit(X_train_sm_sc,  y_train_sm)
y_pred = clf_logregcv_tuned.predict(X_test_sc)
classification(y_test, y_pred, X_test_sc, clf_logregcv_tuned)

Unfortunately, the parameters returned by our grid search did not seem to improve the recall score. This can potentially be due to the limitation of the model itself or more likely is the limitations of our dataset. We simply may not have enough data points to more accurately predict the popularity of a song.

# iNTERPRET

Now that we have 3 tuned models, we can analyze which attributes they used in predicting whether a song was going to be popular or not and interpret these values. For this we will be looking at feature importances of each model and comparing them against each other to see if we can see any common threads between the models.

## Parsing Feature Importances to Dataframes

### Random Forest

In [ ]:
#accessing feature importance values of the tuned random forest model and sorting them
rf_importances_df = pd.Series(clf_rf_tuned.feature_importances_, index=X.columns).sort_values(ascending=False)
#parsing the series to a dataframe
rf_importances_df = rf_importances_df.reset_index()
rf_importances_df.columns = ['RF-Attribute', 'RF-Importance']
rf_importances_df

### XGBoost

In [ ]:
#parsing feature importances to a series and sorting
xgb_importances_df = pd.Series(clf_xgb_tuned.feature_importances_, index=X.columns).sort_values(ascending=False)
#parsing the series to a dataframe
xgb_importances_df = xgb_importances_df.reset_index()
xgb_importances_df.columns=['XGB-Attribute', 'XGB-Importance']
xgb_importances_df

### LogisticRegressionCV

In [ ]:
#accessing feature importance values of the tuned logistic regression model and sorting them
logregcv_importances_df = pd.Series(clf_logregcv_tuned.coef_[0], index=X.columns).sort_values(ascending=False)
#parsing the series to a dataframe
logregcv_importances_df = logregcv_importances_df.reset_index()
logregcv_importances_df.columns = ['LogReg-Attribute', 'LogReg-Importance']
logregcv_importances_df

In [ ]:
#Concatenating feature importances into a single dataframe
importances_df = pd.concat([rf_importances_df, xgb_importances_df, logregcv_importances_df], axis=1)
importances_df

## Feature Importance Comparison

In [ ]:
#plotting feature importances for all models for comparison

fig, ax = plt.subplots(ncols=3, figsize=(15,5))

rf_importances_df = rf_importances_df.sort_values(by='RF-Importance', ascending=True).tail(10)
ax[0].barh(rf_importances_df['RF-Attribute'], rf_importances_df['RF-Importance'])
ax[0].set_title('Feature Importances: Random Forest')

xgb_importances_df = xgb_importances_df.sort_values(by='XGB-Importance', ascending=True).tail(10)
ax[1].barh(xgb_importances_df['XGB-Attribute'], xgb_importances_df['XGB-Importance'])
ax[1].set_title('Feature Importances: XGBoost')

logregcv_importances_df = logregcv_importances_df.sort_values(by='LogReg-Importance', ascending=True).tail(10)
ax[2].barh(logregcv_importances_df['LogReg-Attribute'], logregcv_importances_df['LogReg-Importance'])
ax[2].set_title('Feature Importances: LogisticRegressionCV')
plt.tight_layout()

Among the 3 models we built we can see that Genre of a song has the highest effect on the popularity of a song. On all 3 models, a song having Pop as its genre had the most impact on its popularity. This makes sense since Pop songs by nature are considered popular. Among the rest of the features shown above, different attribute scores such as danceability, energy, different genres and acousticness play a major role. Next, we can inspect the full gamut of the feature importances for logistic regression for reference.

In [ ]:
logregcv_importances_df = pd.Series(clf_logregcv_tuned.coef_[0], index=X.columns).sort_values(ascending=False)
#parsing the series to a dataframe
logregcv_importances_df = logregcv_importances_df.reset_index()
logregcv_importances_df.columns = ['Attribute', 'Importance']

fig, ax = plt.subplots(figsize=(10,15))
ax.barh(logregcv_importances_df['Attribute'], logregcv_importances_df['Importance'])

We can see here that while certain features like 'Pop', 'Rock' and 'danceability' positively affected the prediction, other features such as 'Ska', 'Anime' and 'key_G' negatively affected it. Next we can dive into our processed dataframe and explore some of these attributes for popular and unpopular songs to come to conclusions.

## Data Visualizations

### Genre

In [ ]:
#separating popular and unpopular songs to two dfs
popular_songs_df = df_ohe[df_ohe['is_popular'] == 1]
unpopular_songs_df = df_ohe[df_ohe['is_popular']==0]

In [ ]:
#checking for genre occurence counts for popular songs
popular_genre_df = popular_songs_df.iloc[:, 10:36].agg('sum').sort_values(ascending=False).reset_index()
popular_genre_df.columns = ['genre', 'count']
popular_genre_df

In [ ]:
fig, ax = plt.subplots(figsize=(10, 7))
sns.barplot(x=popular_genre_df['genre'].head(5), y=popular_genre_df['count'].head(5), 
           palette='dark:seagreen_r')

ax=plt.gca()
ax.set_xlabel('Genre')
ax.set_ylabel('Count')
ax.set_title('Top 5 Most Frequent Genres Among Popular Songs')
plt.tight_layout();
plt.savefig('images/genre-popular.jpg')

Above bar graph shows us the most frequent genres among popular songs. As we discussed above, most popular songs have Pop as their genre followed by Rap, Rock, Hip-Hop and Dance. These results make sense and are in-line with a survey conducted by IFPI (https://www.statista.com/chart/15763/most-popular-music-genres-worldwide/).

In [ ]:
#checking for genre occurence counts for unpopular songs
unpopular_genre_df = unpopular_songs_df.iloc[:, 10:36].agg('sum').sort_values(ascending=False).reset_index()
unpopular_genre_df.columns = ['genre', 'count']
unpopular_genre_df

In [ ]:
fig, ax = plt.subplots(figsize=(10,7))
sns.barplot(x=unpopular_genre_df['genre'].head(5), y=unpopular_genre_df['count'].head(5), 
           palette='dark:#5A9_r')

ax=plt.gca()
ax.set_xlabel('Genre')
ax.set_ylabel('Count')
ax.set_title('Top 5 Most Frequent Genres Among Unpopular Songs')
plt.tight_layout();
plt.savefig('images/genre-unpopular.jpg')
# ax.set_xticklabels(ax.get_xticklabels(), rotation=45,ha='center');

The most frequent genres of unpopular songs can be seen above. The results make sense as these genres tend to have a more niche fanbase or as in the case of "Children's Music" are listened to infrequently.

In [ ]:
#displaying percentages for each genre
popular_genre_df['count']=popular_genre_df['count']/popular_genre_df['count'].sum()
popular_genre_df

In [ ]:
#displaying percentages for each genre
unpopular_genre_df['count']=unpopular_genre_df['count']/unpopular_genre_df['count'].sum()
unpopular_genre_df

### Energy

In [ ]:
#removing outliers from energy scores and separating them to Series for popular and unpopular songs
popular_energy_clean = popular_songs_df[find_outliers_IQR(popular_songs_df['energy'])==False]
print(popular_energy_clean['energy'].describe())

unpopular_energy_clean = unpopular_songs_df[find_outliers_IQR(unpopular_songs_df['energy'])==False]
print(unpopular_energy_clean['energy'].describe())

In [ ]:
import numpy as np

#storing mean energy scores in dict
mean_energy = {'popular': popular_energy_clean['energy'].mean(), 
                     'unpopular': unpopular_energy_clean['energy'].mean()}
#visualizing mean scores
with sns.axes_style("whitegrid"):
    fig, ax = plt.subplots(figsize=(8,5))
    ax.barh(y=list(mean_energy.keys()), 
            width=list(mean_energy.values()), 
            color=[sns.color_palette('viridis')[3],sns.color_palette('viridis')[4]]) 
    ax.set_xlim(0, 1)
    ax.set_xticks(np.arange(0,1.1,0.1))
    ax.set_ylabel('Popularity of Songs')
    ax.set_xlabel('Mean Energy Score')
    ax.set_title('Mean Energy Scores for Popular and Unpopular Songs')
    plt.tight_layout()
    plt.savefig('images/energy.jpg')

As we can see above, popular songs tended to be more energetic compared to unpopular songs. This makes sense since the most frequent genres we explored tend to also be energetic genres.

### Danceability

In [ ]:
print('Median Danceability Scores')
print('-------------------')
print(f"Unpopular Songs: {round(unpopular_songs_df['danceability'].median(),2)}")
print(f"Popular Songs: {round(popular_songs_df['danceability'].median(),2)}")

In [ ]:
sns.histplot(data = popular_songs_df, x='danceability', bins='auto')
plt.vlines(x=popular_songs_df['danceability'].median(), ymin=0, ymax=1000, color='red', ls='--')

In [ ]:
sns.histplot(data = unpopular_songs_df, x='danceability', bins='auto')
plt.vlines(x=unpopular_songs_df['danceability'].median(), ymin=0, ymax=4000, color='red', ls='--')

In [ ]:
#removing outliers from danceability scores and separating them to Series for popular and unpopular songs
popular_dance_clean = popular_songs_df[find_outliers_IQR(popular_songs_df['danceability'])==False]
print(popular_dance_clean['danceability'].describe())

unpopular_dance_clean = unpopular_songs_df[find_outliers_IQR(unpopular_songs_df['danceability'])==False]
print(unpopular_dance_clean['danceability'].describe())

In [ ]:
#storing mean danceability scores in dict
mean_danceability = {'popular': popular_dance_clean['danceability'].mean(), 
                     'unpopular': unpopular_dance_clean['danceability'].mean()}

#visualizing mean scores
with sns.axes_style("whitegrid"):
    fig, ax = plt.subplots(figsize=(8,5))
    ax.barh(y=list(mean_danceability.keys()), 
            width=list(mean_danceability.values()), 
            color=[sns.color_palette('viridis')[0],sns.color_palette('viridis')[1]]) 
    ax.set_xlim(0, 1)
    ax.set_xticks(np.arange(0,1.1,0.1))
    ax.set_ylabel('Popularity of Songs')
    ax.set_xlabel('Mean Danceability Score')
    ax.set_title('Mean Danceability Scores for Popular and Unpopular Songs')
    plt.tight_layout();
    plt.savefig('images/danceability.jpg')

Above, it is clear that the popular songs tended to have a higher danceability score compared to unpopular songs. This follows the same trend as the energy scores where majority of the popular songs are high energy and danceable (refer to Appendix A for definition of "danceability": high tempo, high beat strength etc.)

### Acousticness

In [ ]:
#removing outliers from danceability scores and separating them to Series for popular and unpopular songs
popular_acoustic_clean = popular_songs_df[find_outliers_IQR(popular_songs_df['acousticness'])==False]
print(popular_acoustic_clean['acousticness'].describe())

unpopular_acoustic_clean = unpopular_songs_df[find_outliers_IQR(unpopular_songs_df['acousticness'])==False]
print(unpopular_acoustic_clean['acousticness'].describe())

In [ ]:
sns.histplot(data = popular_acoustic_clean, x='acousticness', bins='auto')
plt.vlines(x=popular_acoustic_clean['acousticness'].mean(), ymin=0, ymax=5000, color='red', ls='--')

In [ ]:
sns.histplot(data = unpopular_songs_df, x='acousticness', bins='auto')
plt.vlines(x=unpopular_songs_df['acousticness'].median(), ymin=0, ymax=30000, color='red', ls='--')

In [ ]:
#storing mean acousticness scores in dict
mean_acousticness = {'popular': popular_acoustic_clean['acousticness'].mean(), 
                     'unpopular': unpopular_acoustic_clean['acousticness'].mean()}

#visualizing mean scores
with sns.axes_style("whitegrid"):
    fig, ax = plt.subplots(figsize=(8,5))
    ax.barh(y=list(mean_acousticness.keys()), 
            width=list(mean_acousticness.values()), 
            color=[sns.color_palette('viridis')[2],sns.color_palette('viridis')[3]]) 
    ax.set_xlim(0, 1)
    ax.set_xticks(np.arange(0,1.1,0.1))
    ax.set_ylabel('Popularity of Songs')
    ax.set_xlabel('Mean Acousticness Score')
    ax.set_title('Mean Acousticness Scores for Popular and Unpopular Songs')
    plt.tight_layout();
    plt.savefig('images/acousticness.jpg')

Similar to the energy and danceability scores we see that the popular songs tended to have a lower acousticness score. Since acoustic songs are usually lower energy and rarely danceable this follows the same trend we've been observing.

# CONCLUSIONS & RECOMMENDATIONS

In a competitive environment like the music streaming market, it is vital to retain current subscribers and add new subscribers over time. By accurately predicting which song will be popular next, companies like Spotify can leverage this information to create better playlists and find and sign exclusivity deals with established and up-and-coming artists more easily.  To sum up, our analysis of approximately 176,000 songs from 2019 showed the following:

- Popular songs tend to have Pop, Rap, Rock, Hip-Hop and Dance as their genres.
- More niche genres such as Children's Music, Comedy, Soundtracks, Classical and Jazz tend to be unpopular.
- Generally, popular songs are higher energy, danceable, and therefore less acoustic.